platforms/common/px4_work_queue/WorkQueue.cpp

```diff
@@ -159,6 +161,9 @@ void WorkQueue::Run()
 		}
 
 		work_unlock();
+
+		px4_lockstep_progress(lockstep_component);
+		px4_lockstep_unregister_component(lockstep_component);
```

Comment:
Why do you unregister in each cycle? This will not give any guarantees.

This generally does not work, and I don't think we can have it directly in the work queue unless we know it's scheduled on each cycle.

Reply:
The only guarantees are
 1. for each new gyro sample there's a new actuator output
 2. any scheduled WorkItem (including WorkItems scheduled by WorkItems) is allowed to run to completion

> Why do you unregister in each cycle? 

Because we only care about something scheduled that step running to completion. Every WorkQueue isn't required to cycle every step.

Reply:
That's exactly the problem: triggering a schedule does not mean it's woken up in the same cycle and thus you lose all guarantees. It's purely up to the scheduling whether this works.